tests/unit/s2n_tls13_handshake_test.c

```diff
@@ -28,315 +28,455 @@
-        tls13_state_machine[i].handler[0] = s2n_test_handler;
-        tls13_state_machine[i].handler[1] = s2n_test_handler;
-    }
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
 
-    /* Test: When using TLS 1.3, use the new state machine and handshakes */
-    {
-        struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT);
-        conn->actual_protocol_version = S2N_TLS13;
-        EXPECT_EQUAL(ACTIVE_STATE_MACHINE(conn), tls13_state_machine);
-        EXPECT_EQUAL(ACTIVE_HANDSHAKES(conn), tls13_handshakes);
-        EXPECT_SUCCESS(s2n_connection_free(conn));
-    }
+        client_conn->actual_protocol_version = S2N_TLS13;
+        server_conn->actual_protocol_version = S2N_TLS13;
 
-    /* Test: TLS1.3 server does not wait for client cipher change requests */
-    {
-        struct s2n_connection *conn = s2n_connection_new(S2N_SERVER);
-        conn->actual_protocol_version = S2N_TLS13;
+        struct s2n_stuffer client_hello_key_share;
+        struct s2n_stuffer server_hello_key_share;
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&client_hello_key_share, 1024));
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&server_hello_key_share, 1024));
 
-        for (int i = 0; i < valid_tls13_handshakes_size; i++) {
-            int handshake = valid_tls13_handshakes[i];
+        /* Client sends ClientHello key_share */
+        EXPECT_SUCCESS(s2n_extensions_client_key_share_send(client_conn, &client_hello_key_share));
+        S2N_STUFFER_READ_EXPECT_EQUAL(&client_hello_key_share, TLS_EXTENSION_KEY_SHARE, uint16);
+        S2N_STUFFER_READ_EXPECT_EQUAL(&client_hello_key_share, s2n_extensions_client_key_share_size(server_conn) - 4, uint16);
 
-            conn->handshake.handshake_type = handshake;
+        EXPECT_SUCCESS(s2n_extensions_client_key_share_recv(server_conn, &client_hello_key_share));
 
-            for (int j = 0; j < S2N_MAX_HANDSHAKE_LENGTH; j++) {
-                if (tls13_handshakes[i][j] == CLIENT_CHANGE_CIPHER_SPEC) {
-                    conn->handshake.message_number = j - 1;
+        /* Server configures the "negotiated_curve" */
+        server_conn->secure.server_ecc_params.negotiated_curve = &s2n_ecc_supported_curves[0];
 
-                    EXPECT_SUCCESS(s2n_advance_message(conn));
+        /* Server sends ServerHello key_share */
+        EXPECT_SUCCESS(s2n_extensions_server_key_share_send(server_conn, &server_hello_key_share));
 
-                    EXPECT_EQUAL(conn->handshake.message_number, j + 1);
-                    EXPECT_NOT_EQUAL(ACTIVE_MESSAGE(conn), CLIENT_CHANGE_CIPHER_SPEC);
+        S2N_STUFFER_READ_EXPECT_EQUAL(&server_hello_key_share, TLS_EXTENSION_KEY_SHARE, uint16);
+        S2N_STUFFER_READ_EXPECT_EQUAL(&server_hello_key_share, s2n_extensions_server_key_share_send_size(server_conn) - 4, uint16);
```

Comment:
We could split this into:

```
EXPECT_EQUAL(0, s2n_extensions_server_key_share_send_size(server_conn));
S2N_STUFFER_READ_EXPECT_EQUAL(&server_hello_key_share, s2n_extensions_server_key_share_send_size(server_conn) - 4, uint16);
```

and why is this server_key_share_size `-4?` Is it beacuse of `- (S2N_SIZE_OF_EXTENSION_TYPE + S2N_SIZE_OF_EXTENSION_DATA_SIZE)` if so we should use the latter to be clear.

Reply:
s2n_extensions_server_key_share_send_size(server_conn) actually gives `key_share_size + S2N_SIZE_OF_EXTENSION_TYPE + S2N_SIZE_OF_EXTENSION_DATA_SIZE`. the detailed tests are in `key_share_extension_test` so they aren't necessary for this symmetric test. I'll update the constants for clarity